src/llm_tracekit/gemini/patch.py

```diff
@@ -0,0 +1,480 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from time import perf_counter_ns
+from typing import Any, AsyncIterator, Iterator, Optional
+
+from opentelemetry.semconv._incubating.attributes import (
+    gen_ai_attributes as GenAIAttributes,
+)
+from opentelemetry.trace import SpanKind, Tracer
+
+from llm_tracekit.gemini.state import GeminiOperationState, GeminiSpanContext
+from llm_tracekit.gemini.utils import (
+    build_request_details,
+    build_response_details,
+)
+from llm_tracekit.instrumentation_utils import handle_span_exception
+from llm_tracekit.instruments import Instruments
+
+
+_GEMINI_SYSTEM_VALUE = getattr(
+    GenAIAttributes.GenAiSystemValues,
+    "GOOGLE_GENAI",
+    "google_genai",
+)
+
+
+@dataclass
+class _WrapperConfig:
+    tracer: Tracer
+    instruments: Instruments
+    capture_content: bool
+
+
+def generate_content_wrapper(
+    tracer: Tracer,
+    instruments: Instruments,
+    capture_content: bool,
+):
+    config = _WrapperConfig(tracer=tracer, instruments=instruments, capture_content=capture_content)
+
+    def traced_method(wrapped, instance, args, kwargs):
+        model = _get_argument(args, kwargs, name="model", position=0)
+        contents = _get_argument(args, kwargs, name="contents", position=1)
+        system_instruction = _get_argument(args, kwargs, name="system_instruction")
+        config_payload = _get_argument(args, kwargs, name="config", position=2)
+
+        request_details = build_request_details(
+            model=model,
+            contents=contents,
+            system_instruction=system_instruction,
+            config=config_payload,
+            capture_content=config.capture_content,
+        )
+
+        with config.tracer.start_as_current_span(
+            name=request_details.span_name,
+            kind=SpanKind.CLIENT,
+            attributes=request_details.span_attributes,
+            end_on_exit=False,
+        ) as span:
+            operation_state = _prepare_operation_state(span, request_details, config.capture_content)
+
+            try:
+                result = wrapped(*args, **kwargs)
+                operation_state.response_details = build_response_details(
+                    response=result,
+                    capture_content=config.capture_content,
+                )
+                operation_state.finish_reasons = operation_state.response_details.finish_reasons
+
+                if span.is_recording():
+                    span.set_attributes(operation_state.response_details.span_attributes)
+
+                span.end()
```

Comment:
why do you end the span manually? I'm pretty sure in case of error we still want to end the span